lib/python/picongpu/picmi/simulation.py

```diff
@@ -28,6 +28,11 @@ class Simulation(picmistandard.PICMI_Simulation):
     https://picmi-standard.github.io/standard/simulation.html
     """
 
+    picongpu_custom_input = util.build_typesafe_property(
+        typing.Optional[list[customuserinput.CustomUserInput | rendering.RenderedObject]]
+    )
```

Comment:
Damn! That's some dark python magic. Didn't know that.

Reply:
well this particular version is a mixture,
- `util.build_typesafe_property(<type>)` is picongpu specific type and assigment enforcement
- the type declaration is a python language feature, its just not usually enforced unless you use external decorators or enforce it yourself

Reply:
Yeah, I meant what `build_typesafe_property` does. The other one is standard Python (although admittedly I think I've never used the classes from `typing` as objects yet).